lib/datadog/grok/src/parse_grok_rules.rs

```diff
@@ -0,0 +1,375 @@
+        &mut prev_rule_filters,
+        &mut grok,
+    )?;
+    // parse match rules and return them
+    parse_rules(
+        parsing_rules,
+        &mut prev_rule_definitions,
+        &mut prev_rule_filters,
+        &mut grok,
+    )
+}
+
+fn parse_rules<'a>(
+    parsing_rules: &'a [String],
+    mut prev_rule_definitions: &mut HashMap<&'a str, String>,
+    mut prev_rule_filters: &mut HashMap<&'a str, HashMap<LookupBuf, Vec<GrokFilter>>>,
+    mut grok: &mut Grok,
+) -> Result<Vec<GrokRule>, Error> {
+    parsing_rules
+        .iter()
+        .filter(|&r| !r.is_empty())
+        .map(|r| {
+            parse_grok_rule(
+                r,
+                &mut prev_rule_definitions,
+                &mut prev_rule_filters,
+                &mut grok,
+            )
+        })
+        .collect::<Result<Vec<GrokRule>, Error>>()
+}
+
+fn parse_grok_rule<'a>(
+    rule: &'a str,
+    mut prev_rule_definitions: &mut HashMap<&'a str, String>,
+    mut prev_rule_filters: &mut HashMap<&'a str, HashMap<LookupBuf, Vec<GrokFilter>>>,
+    grok: &mut Grok,
+) -> Result<GrokRule, Error> {
+    let mut split_whitespace = rule.splitn(2, ' ');
+    let split = split_whitespace.by_ref();
+    let rule_name = split.next().ok_or_else(|| {
+        Error::InvalidGrokExpression(
+            rule.to_string(),
+            "format must be: 'ruleName definition'".into(),
+        )
+    })?;
+    let mut rule_def = split
+        .next()
+        .ok_or_else(|| {
+            Error::InvalidGrokExpression(
+                rule.to_string(),
+                "format must be: 'ruleName definition'".into(),
+            )
+        })?
+        .to_string();
+
+    let rule_def_cloned = rule_def.clone();
+    lazy_static! {
+        static ref GROK_PATTER_RE: Regex = Regex::new(r"%\{.+?\}").unwrap();
+    }
+    // find all patterns %{}
+    let raw_grok_patterns = GROK_PATTER_RE
+        .find_iter(rule_def_cloned.as_str())
+        .map(|rule| rule.as_str())
+        .collect::<Vec<&str>>();
+    // parse them
+    let mut grok_patterns = raw_grok_patterns
+        .iter()
+        .map(|pattern| {
+            parse_grok_pattern(pattern)
+                .map_err(|e| Error::InvalidGrokExpression(pattern.to_string(), e.to_string()))
+        })
+        .collect::<Result<Vec<GrokPattern>, Error>>()?;
+
+    grok_patterns = index_repeated_fields(grok_patterns);
+
+    let mut filters: HashMap<LookupBuf, Vec<GrokFilter>> = HashMap::new();
+    let pure_grok_patterns: Vec<String> = grok_patterns
+        .iter()
+        .map(|pattern| {
+            purify_grok_pattern(
+                &pattern,
+                &mut filters,
+                &mut prev_rule_filters,
+                &mut prev_rule_definitions,
+            )
+        })
+        .collect::<Result<Vec<String>, Error>>()?;
+
+    // replace grok patterns with "purified" ones
+    let mut pure_pattern_it = pure_grok_patterns.iter();
+    for r in raw_grok_patterns {
```

Comment:
```suggestion
    for (r, pure) in raw_grok_patterns.iter().zip(pure_grok_patterns.iter()) {
```

Reply:
Fixed in https://github.com/timberio/vector/pull/8850/commits/d924c5370e4f02e5fb595b7aa006870a75aa6a70